datasize/src/std.rs

```diff
@@ -198,27 +198,46 @@ where
     }
 }
 
+fn estimate_hashbrown_rawtable<T>(capacity: usize) -> usize {
+    // https://github.com/rust-lang/hashbrown/blob/v0.12.3/src/raw/mod.rs#L185
+    let buckets = if capacity < 8 {
+        if capacity < 4 {
+            4
+        } else {
+            8
+        }
+    } else {
+        (capacity * 8 / 7).next_power_of_two()
+    };
+    // https://github.com/rust-lang/hashbrown/blob/v0.12.3/src/raw/mod.rs#L242
+    let size = size_of::<T>();
+    // `Group` is u32, u64, or __m128i depending on the CPU architecture.
+    // Return a lower bound, ignoring its constant contributions
+    // (through ctrl_align and Group::WIDTH, at most 31 bytes).
+    let ctrl_offset = size * buckets;
+    ctrl_offset + buckets
+}
+
 impl<K, V, S> DataSize for std::collections::HashMap<K, V, S>
 where
     K: DataSize,
     V: DataSize,
 {
-    // Approximation directly taken from
-    // https://github.com/servo/heapsize/blob/f565dda63cc12c2a088bc9974a1b584cddec4382/src/lib.rs#L266-L275
     const IS_DYNAMIC: bool = true;
 
     const STATIC_HEAP_SIZE: usize = 0;
 
     #[inline]
     fn estimate_heap_size(&self) -> usize {
-        let size = self.capacity() * (size_of::<V>() + size_of::<K>() + size_of::<usize>());
+        let size = estimate_hashbrown_rawtable::<(K, V)>(self.capacity());
 
         if K::IS_DYNAMIC || V::IS_DYNAMIC {
-            self.iter().fold(size, |n, (key, value)| {
-                n + key.estimate_heap_size() + value.estimate_heap_size()
-            })
+            size + self
+                .iter()
+                .map(|(k, v)| k.estimate_heap_size() + v.estimate_heap_size())
+                .sum::<usize>()
         } else {
-            size + self.capacity() * (K::STATIC_HEAP_SIZE + V::STATIC_HEAP_SIZE)
+            size + self.len() * (K::STATIC_HEAP_SIZE + V::STATIC_HEAP_SIZE)
```

Comment:
I'm not 100% sure how this works, is there no preallocation happening and thus no need to consider `capacity`?

Reply:
`estimate_hashbrown_rawtable` already accounts for the allocation made directly by hashbrown, based on capacity. Everything after that is allocations owned by the keys and values in the map, so empty hashtable buckets don’t have any.

Another way to look at this is that the `else` is an optimization that should give the same result as the other `if` branch, and `self.iter().len() == self.len()`.